app/ide-desktop/lib/client/src/index.ts

```diff
@@ -213,6 +219,55 @@ class App {
             })
         }
         electron.ipcMain.on(ipc.channel.quit, () => electron.app.quit())
+        // Register an IPC listener for the `openExternalUrl` event. This event is emitted by the
+        // authentication package when it needs to open a URL in the user's default (i.e., system)
+        // browser. This is used because we don't want to run the OAuth flow in the app, because
+        // users don't trust Electron apps to handle their credentials.
+        electron.ipcMain.on(ipc.channel.openExternalUrl, (_event, url) => opener(url))
+    }
+
+    /**
+     * Initialize the listener for `open-url` events.
+     *
+     * This listener is used to open a page in *this* application window, when the user is
+     * redirected to a URL with a protocol supported by this application.
+     *
+     * For example, when the user completes an OAuth sign in flow (e.g., through Google), they are
+     * redirected to a URL like `enso://auth?code=...`. This listener will intercept that URL
+     * and open the page `auth?code=...` in the application window.
+     */
+    initOpenUrlListener() {
+        electron.app.on('open-url', (event, url) => {
+            // We only want to handle URLs that are both deep links and that are meant for this
+            // application. We can tell if it is by checking the protocol. Any non-matching URLs are
+            // ignored by this handler (and passed to the default handler). Any matchin URLs are
+            // handled **only** by this handler.
+            const parsedUrl = new URL(url)
+            const isDeepLink = parsedUrl.protocol === `${DEEP_LINK_PROTOCOL}:`
+            if (!isDeepLink) {
+                return
+            }
+
+            // Don't open the deep link URL in the window, we want the system browser to handle it.
+            event.preventDefault()
+
+            // Parse the relevant details from the URL, and pass it to the designated authentication
+            // URL handler over the IPC channel. This URL handler should have been registered by the
+            // `preload` module, and is responsible for opening the URL in the system browser.
+            //
+            // TODO [NP]: https://github.com/enso-org/cloud-v2/issues/339
+            // The reason we parse and re-create the URL here at all is because some of the URLs
+            // sent by our authentication flows are malformed. That is, some take the form
+            // `enso://localhost:8080/auth` and others take the form `enso://auth`. What we are
+            // doing here is effectively normalizing the URLs to always start with a known value (in
+            // this case) `http://localhost:8080` instead. This is a hack, and should be fixed by
+            // making the authentication flows send consistent URLs in the first place.
```

Comment:
This is not a bug. We use inconsistent hostnames in the Cognito email templates, so our parsing logic is more complex than it could be.

URL must contain *something* for the hostname, so that we can parse the URL. But it doesn't matter what it is because we are interested in only the scheme + path + search, not the hostname.

Some of our URLs are missing the hostname entirely. So this comment is suggesting we use `localhost:8080` as the hostname for all of them, to be consistent.

We could choose something arbitrary like `authentication` for the hostname instead. For example: `enso://authentication/registration`.